libraries/lib-graphics/graphics/Rect.h

```diff
@@ -0,0 +1,267 @@
+   PointType<DataType> origin;
+   //! Size of the rectangle.
+   SizeType<DataType> size;
+
+   //! Return true if the rectangle is not empty.
+   bool IsValid() const noexcept
+   {
+      return size.width > 0 && size.height > 0;
+   }
+
+   //! Return true if the rectangle contains the point specified.
+   bool Contains(PointType<DataType> pt) const noexcept
+   {
+      return pt.x >= origin.x && pt.y >= origin.y &&
+             (pt.x < (origin.x + size.width)) &&
+             (pt.y < (origin.y + size.height));
+   }
+};
+
+//! Calculates the intersection of two rectangles.
+template <typename DataType>
+RectType<DataType>
+Intersect(RectType<DataType> firstRect, RectType<DataType> secondRect) noexcept
+{
+   if (firstRect.size.IsZero() || secondRect.size.IsZero())
+      return {};
+
+   const PointType<DataType> p11 = firstRect.origin;
+   const PointType<DataType> p12 = { firstRect.origin.x + firstRect.size.width,
+                                     firstRect.origin.y +
+                                        firstRect.size.height };
+   const PointType<DataType> p21 = secondRect.origin;
+   const PointType<DataType> p22 = {
+      secondRect.origin.x + secondRect.size.width,
+      secondRect.origin.y + secondRect.size.height
+   };
+
+   if (p12.x < p21.x || p11.x > p22.x || p11.y > p22.y || p12.y < p21.y)
+      return {};
+
+   const PointType<DataType> p31 = { std::max(p11.x, p21.x),
+                                     std::max(p11.y, p21.y) };
+
+   const PointType<DataType> p32 = { std::min(p12.x, p22.x),
+                                     std::min(p12.y, p22.y) };
+
+   return { p31, SizeType<DataType> { p32.x - p31.x, p32.y - p32.y - p31.y } };
+}
+
+template <typename DataType>
+bool operator==(
+   const RectType<DataType> lhs, const RectType<DataType> rhs) noexcept
+{
+   return lhs.origin == rhs.origin && lhs.size == rhs.size;
+}
+
+template <typename DataType>
+bool operator!=(
+   const RectType<DataType> lhs, const RectType<DataType> rhs) noexcept
+{
+   return !(lhs == rhs);
+}
+
+//! Cast the rectangle to another data type.
+template <typename To, typename From>
+RectType<To> rect_cast(RectType<From> rect)
+{
+   return { point_cast<To>(rect.origin), size_cast<To>(rect.size) };
+}
+
+//! Alias for rectangle with float data type.
+using Rect = RectType<float>;
+
+//! Axis aligned bounding box.
+template<typename DataType>
+struct AxisAlignedBoundingBoxType
+{
+   //! Top left corner of the box
+   PointType<DataType> topLeft { std::numeric_limits<DataType>::max(),
+                                 std::numeric_limits<DataType>::max() };
+
+   //! Bottom right corner of the box
+   PointType<DataType> bottomRight { std::numeric_limits<DataType>::min(),
+                                     std::numeric_limits<DataType>::min() };
+
+   //! Return true if the box is not empty.
+   bool IsValid() const noexcept
+   {
+      return topLeft.x <= bottomRight.x && topLeft.y <= bottomRight.y;
```

Comment:
So the convention is that x coordinates increase rightwards, y coordinates downwards.  It's the usual graphics convention, but worth explicit comment near the top of the class definition.

Reply:
A box could be valid but empty, if bottom and right are excluded.